core/lib/types/src/fee_model.rs

```diff
@@ -0,0 +1,179 @@
+                fair_pubdata_price: input.l1_gas_price * 17,
+                l1_gas_price: input.l1_gas_price,
+            },
+        }
+    }
+
+    pub fn for_protocol_version(
+        protocol_version: ProtocolVersionId,
+        fair_l2_gas_price: u64,
+        fair_pubdata_price: Option<u64>,
+        l1_gas_price: u64,
+    ) -> Self {
+        if protocol_version.is_1_4_1() {
+            Self::PubdataIndependent(PubdataIndependentBatchFeeModelInput {
+                fair_l2_gas_price,
+                fair_pubdata_price: fair_pubdata_price
+                    .expect("Pubdata price must be provided for protocol version 1.4.1"),
+                l1_gas_price,
+            })
+        } else {
+            Self::L1Pegged(L1PeggedBatchFeeModelInput {
+                fair_l2_gas_price,
+                l1_gas_price,
+            })
+        }
+    }
+}
+
+/// Pubdata is only published via calldata and so its price is pegged to the L1 gas price.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct L1PeggedBatchFeeModelInput {
+    /// Fair L2 gas price to provide
+    pub fair_l2_gas_price: u64,
+    /// The L1 gas price to provide to the VM.
+    pub l1_gas_price: u64,
+}
+
+/// Pubdata price may be independent from L1 gas price. The L1 gas price is not needed.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PubdataIndependentBatchFeeModelInput {
+    /// Fair L2 gas price to provide
+    pub fair_l2_gas_price: u64,
+    /// Fair pubdata price to provide.
+    /// In this version, it MUST be equal to 17 * l1_gas_price
+    pub fair_pubdata_price: u64,
+    /// The L1 gas price to provide to the VM. Even if some of the VM versions may not use this value, it is still maintained for backward compatibility.
+    pub l1_gas_price: u64,
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct MainNodeFeeModelConfig {
+    /// The factor by which the L1 gas price is scaled. This is used to account for the fact that the L1 gas price may fluctuate.
+    pub l1_gas_price_scale_factor: f64,
+    /// The factor by which the L1 pubdata price is scaled. This is used to account for the fact that the L1 pubdata price may fluctuate.
+    pub l1_pubdata_price_scale_factor: f64,
+    /// The minimal acceptable L2 gas price, i.e. the price that should include the cost of computation/proving as well
+    /// as potentially premium for congestion.
+    pub minimal_l2_gas_price: u64,
+    /// Tthe constant that represents the possibility that a batch can be sealed because of overuse of compute.
+    pub compute_overhead_percent: f64,
+    /// The constant that represents the possibility that a batch can be sealed because of overuse of pubdata.
+    pub pubdata_overhead_percent: f64,
+    /// The constant amount of L1 gas that is used as the overhead for the batch. It includes the price for batch verification, etc.
+    pub batch_overhead_l1_gas: u64,
+    /// The maximum amount of gas that can be used by the batch.
```

Comment:
mention what are the natural limitations - why people should restrict it and not set to 100B